### ISSUE TITLE
Support for chai v5

### ISSUE THREAD

**Issue by attekemppila** [prodatakey/dirty-chai#54]
node v18.19.0
npm v10.2.3

Current peerDependency doesn't include [chai](https://github.com/chaijs/chai) v5:

https://github.com/prodatakey/dirty-chai/blob/bb5d007b40a57080205afb81f7405cd1e8b89e84/package.json#L39-L41

npm install fails because of that.

Does anything else need chaning in the code. I didn't test yet myself.

**Comment by perrin4869** [comment]
I started looking into this, the main problem right now is that we cannot monkey-patch the utils object anymore, since it is a [non-configurable getter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty#configurable), so in chai v5 [this](https://github.com/prodatakey/dirty-chai/blob/bb5d007b40a57080205afb81f7405cd1e8b89e84/lib/dirty-chai.js#L170) will throw.
The solution is to remove the 3 money patches and expect users to install `dirty-chai` as the last plugin, or to make the getter `configurable` in `chai@5`. 
@keithamus any opinions?
@joshperry do you still have write permissions to this repo?

**Comment by keithamus** [comment]
Happy to make getters configurable in chai. Please file an issue with the explanation on the chai issue tracker so we can track the work there (or better yet, please file a PR :wink:).

**Comment by lehni** [comment]
I just looked into it. These getters are not produced by chai's own code, but by the `esbuild` command. I don't think there's an easy way to make them configurable through `esbuild`:

```js
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// (disabled):util
var require_util = __commonJS({
  "(disabled):util"() {
  }
});

// lib/chai/utils/index.js
var utils_exports = {};
__export(utils_exports, {
  addChainableMethod: () => addChainableMethod,
  addLengthGuard: () => addLengthGuard,
  addMethod: () => addMethod,
  addProperty: () => addProperty,
  checkError: () => check_error_exports,
  compareByInspect: () => compareByInspect,
  eql: () => deep_eql_default,
  expectTypes: () => expectTypes,
  flag: () => flag,
  getActual: () => getActual,
  getMessage: () => getMessage2,
  getName: () => getName,
  getOperator: () => getOperator,
  getOwnEnumerableProperties: () => getOwnEnumerableProperties,
  getOwnEnumerablePropertySymbols: () => getOwnEnumerablePropertySymbols,
  getPathInfo: () => getPathInfo,
  hasProperty: () => hasProperty,
  inspect: () => inspect2,
  isNaN: () => isNaN2,
  isProxyEnabled: () => isProxyEnabled,
  isRegExp: () => isRegExp2,
  objDisplay: () => objDisplay,
  overwriteChainableMethod: () => overwriteChainableMethod,
  overwriteMethod: () => overwriteMethod,
  overwriteProperty: () => overwriteProperty,
  proxify: () => proxify,
  test: () => test,
  transferFlags: () => transferFlags,
  type: () => type
});

```

**Comment by keithamus** [comment]
We could also look into providing hooks into this to allow plugins like this one to do what it does, perhaps an `onChainableMethodAdded()` function or something?

**Comment by lehni** [comment]
Yes I was thinking about this too. It would need to cover these hooks:

https://github.com/prodatakey/dirty-chai/blob/bb5d007b40a57080205afb81f7405cd1e8b89e84/lib/dirty-chai.js#L168-L214

**Comment by LucaFranceschini** [comment]
Is there any plan for v5 support?

**Comment by perrin4869** [comment]
@joshperry nudge, in case you still have access to this repo?

**Comment by joshperry** [comment]
> [@joshperry](https://github.com/joshperry) nudge, in case you still have access to this repo?

Just gave the guys at PDK access to publish new versions on npm as they are interested in reving support! I'll let them pitch in when they're ready. If you have any changes that would help I'm sure they'd be appreciated.

**Comment by perrin4869** [comment]
oh wow that's great news!! thanks!

**Comment by malwarebo** [comment]
Any update on this issue?

**Comment by perrin4869** [comment]
@joshperry just put together a PR that targets this issue, hopefully it can finally be closed soon :)

**Comment by joshperry** [comment]
Thanks for doing that work @perrin4869! I've published this to npm as [`3.0.0-rc.1`](https://www.npmjs.com/package/dirty-chai/v/3.0.0-rc.1).

I don't have any projects using this right now, but I'm happy to facilitate any way I can.

**Comment by perrin4869** [comment]
thanks for the prompt publish! I just tried it out on one of my good old npm modules that uses `dirty-chai` and was stuck on `chai@4` and it seems to work: https://github.com/dotcore64/gulp-xml-transformer/pull/295

**Comment by yelworc** [comment]
Thanks so much for this! I just updated a project with >2k tests to chai@6 and dirty-chai@3.0.0-rc.1, and everything seems to still be working well. That project is using two other chai plugins, too (just in case that's relevant).

**Comment by attekemppila** [comment]
Thank you @perrin4869 . I tested chai@6.2.0 and dirty-chai@3.0.0-rc.1 in my project (~3k test; no other 3rd party plugins than dirty-chai; a few local custom chai helpers). Everything seemed to work fine.

**Comment by joshperry** [comment]
Pushed this as v3.0.0, seems the testing done so far for getting out of RC is fine for a major bump (nobody will run it without knowing what they're doing).
